app/models/model.rb

```diff
@@ -4,6 +4,13 @@ class Model < ApplicationRecord
   extend FriendlyId
   friendly_id :name
   include Parameterizable::ByIdName
+  include ::Foreman::EventSubscribers::Observable
+
+  notify_event_observers on: :create, with: :created
+  notify_event_observers on: :update, with: :updated
+  notify_event_observers on: :destroy, with: :destroyed do |model|
+    { id: model.id, name: model.name }
+  end
```

Comment:
Looks good, maybe changing the DSL method name to something more readable but I like it how it is designed. One nitpick tho - the model name is automatically deduced which is nice for coder, however if we would explicitly set it here via parameter it's not only more clear for everyone, it would also be possible to grep whole codebase and easily extract list of all hooks which are available. Something which is really painful today for foreman_hooks - there is no list/documentation of hooks available and users need to find them really hard. Alternatively, you can create some registry and a rake task that would list them all.

In fact, I am so confident in this design that I would not hesitate to put this into ApplicationRecord class to be available for all models. It's pretty safe now and if we solve to "discoverability" of those this could be it. However explicitly listing them is also a solid option.

Reply:
Thinking more about this, I remember that AR STI was causing issues to "generic" approach, therefore I lean towards being more explicit and literally requiring all observer point to explicitly list:

* hook name
* hook action
* payload

We can then decide what to do with Subnets or Hosts where there is STI and makes things more complex. When hook registry is also implemented, we can add rake task to list them and also get rid of walking down the ActiveRecord descendants tree.

Reply:
What do you guys think about this:

We call them `lifecycle hooks` if they react to model changes (create, delete, update) or simply `hooks` if a developer needs to fire them manually?

Reply:
I see no reason to differentiate between them from users perspective.